### PR TITLE
Port PX4 to BeagleBone Blue Board using library librobotcontrol instead of a submodule

### PR DESCRIPTION
**Pull request by UAV-Pilot**
James Strawson, the creator of https://github.com/StrawsonDesign/librobotcontrol, helped to make some changes to facilitate the change of package usage in this porting from a submodule to a library.

Corresponding DriverFramework pull request (https://github.com/PX4/DriverFramework/pull/228) was also updated to use librobotcontrol in form of a library instead of a submodule. 

The porting code is constructed in the way that it won't break builds of other configuration targets while allow build of BeagleBone Blue target with some manual setup which is not complicated.

It seems that currently librobotcontrol debian package is only available on BeagleBoard products. Here are the ways to obtain the librobotcontrol on BeagleBone Blue (and Black):
1) use the one pre-installed in  BeagleBoard images
2) install from debian package or repository (sudo apt update && sudo apt install librobotcontrol)
3) install from source
    git clone https://github.com/StrawsonDesign/librobotcontrol.git
    cd librobotcontrol
    sudo make install

After acquiring the pre-built library, 

     a) select and define LIBROBOTCONTROL_INSTALL_DIR environment variable so that other unwanted headers will not be included

     b) install robotcontrol.h and rc/* into $LIBROBOTCONTROL_INSTALL_DIR/include

     c) install pre-built native (ARM) version of librobotcontrol.* into $LIBROBOTCONTROL_INSTALL_DIR/lib

then the BeagleBone Blue target can be built on both cross compile host system and native build system, e.g., make posix_bbblue_cross [upload | upload_px4].

### CODE REVIEW

File: src/modules/mavlink/mavlink_main.cpp [hunk 1]
```diff
@@ -1926,6 +1932,10 @@ Mavlink::task_main(int argc, char *argv[])
 	_mode = MAVLINK_MODE_NORMAL;
 	bool _force_flow_control = false;
 
+#ifdef __PX4_POSIX_BBBLUE
+	_mavlink_wifi_name = __PX4_BBBLUE_DEFAULT_MAVLINK_WIFI;
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.cpp]: Can we make this more generic and remove the ifdefs and avoid the board-specific init?
`_mavlink_wifi_name` -> `_network_interface_name`

File: src/modules/sensors/sensor_params_battery.c [hunk 2]
```diff
@@ -98,10 +98,20 @@ PARAM_DEFINE_FLOAT(BAT_A_PER_V, -1.0);
  * means that measurements are expected to come from a power module. If the value is set to
  * 'External' then the system expects to receive mavlink battery status messages.
  *
- * @min 0
+ * @min -1
```
Comment (bkueng) [inline src/modules/sensors/sensor_params_battery.c]: Why is this changed?

File: platforms/posix/cmake/px4_impl_os.cmake [hunk 3]
```diff
@@ -283,6 +283,20 @@ function(px4_os_add_flags)
...
+	elseif ("${BOARD}" STREQUAL "bbblue")
+		set(BBBLUE_COMPILE_FLAGS -mcpu=cortex-a8 -mfpu=neon -mfloat-abi=hard -mtune=cortex-a8)
+		list(APPEND added_c_flags   ${BBBLUE_COMPILE_FLAGS})
+		list(APPEND added_cxx_flags ${BBBLUE_COMPILE_FLAGS})
+
+		# TODO: Wmissing-field-initializers ignored on older toolchain, can be removed eventually
+		# On cross compile host system and native build system:
+		#   a) install robotcontrol.h and rc/* into /usr/local/include
+		#   b) install pre-built native (ARM) version of librobotcontrol.* into /usr/local/lib
+		list(APPEND added_cxx_flags -I/usr/local/include -Wno-missing-field-initializers)
```
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply platforms/posix/cmake/px4_impl_os.cmake]: Indeed I encountered problem when including other directories. Change to using ENV variable LIBROBOTCONTROL_INSTALL_DIR

File: platforms/posix/src/main.cpp [hunk 4]
```diff
@@ -366,6 +368,9 @@ int main(int argc, char **argv)
 				cout << "node name: " << node_name << endl;
 			}
 
+		} else if (strchr(argv[index], '>')) {
```
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply platforms/posix/src/main.cpp]: What actually encountered in Debian on BBBlue was that the output redirect was passed to argv.

File: posix-configs/bbblue/px4.config [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+#rgbled start -b 1
+bbblue_adc start
+bbblue_adc test
+
+sensors start
+commander start
+navigator start
+ekf2 start
+#land_detector start multicopter
```
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply posix-configs/bbblue/px4.config]: I did not think there was any land detection hardware in my test quadcopter, so I commented it out. Since this is in a configuration file, user can change it accordingly.

File: src/drivers/linux_pwm_out/bbblue_pwm_rc.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+#ifdef __DF_BBBLUE
```
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply src/drivers/linux_pwm_out/bbblue_pwm_rc.cpp]: Changed to __PX4_POSIX_BBBLUE

File: src/drivers/linux_sbus/linux_sbus.cpp [hunk 7]
```diff
@@ -42,7 +42,8 @@ int RcInput::init()
 	/**
 	 * initialize the data of each channel
 	 */
-	for (i = 0; i < input_rc_s::RC_INPUT_MAX_CHANNELS; ++i) {
+	for (i = 0; i < input_rc_s::RC_INPUT_MAX_CHANNELS && i < 16; ++i) {
+		//_data.values defined as int[16] and RC_INPUT_MAX_CHANNELS (=18) > 16
```
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply src/drivers/linux_sbus/linux_sbus.cpp]: Rolled back. Since it's defined in a auto-generated file, I cannot figure out why it's an array of size 16 before.

File: src/modules/sensors/voted_sensors_update.cpp [hunk 8]
```diff
@@ -626,6 +626,7 @@ void VotedSensorsUpdate::accel_poll(struct sensor_combined_s &raw)
 
 	// write the best sensor data to the output variables
 	if (best_index >= 0) {
+		raw.timestamp = _last_sensor_data[best_index].timestamp;
```
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply src/modules/sensors/voted_sensors_update.cpp]: Removed. The problem I had was caused by error_count field not set (refer to change in df_...._wrapper.cpp) , but I over reacted by also attempting to make sure timestamp fields is set.

File: src/modules/mavlink/mavlink_main.cpp (hunk 1 above)
Comment (UAV-Pilot) [reply src/modules/mavlink/mavlink_main.cpp]: Made it generic.

File: src/modules/sensors/sensor_params_battery.c (hunk 2 above)
Comment (UAV-Pilot) [reply src/modules/sensors/sensor_params_battery.c]: Rolled back. I thought -1 was the default value.

File: src/drivers/boards/bbblue/init.c [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+		PX4_ERR("rc_init failed to run rc_servo_power_rail_en(1)");
+		return -1;
+	}
+
+	//i2c, barometer and mpu will be initialized later
+
+	rc_set_state(RUNNING);
+
+	// wait to let threads start up
+	rc_usleep(10000);
```
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply src/drivers/boards/bbblue/init.c]: removed the sleep

**Comment by UAV-Pilot** [comment]
Before commit 2e1399d, all checks were passed. The changes in the last 2 commits ( 2e1399d and 651ba6b ) do not look like to be the cause of flash overflow issue in px4fmu-* builds. Other pull requests had similar problems with px4fmu-*.

File: src/modules/mavlink/mavlink_main.h [hunk 10]
```diff
@@ -601,6 +605,9 @@ class Mavlink
 	uint8_t _network_buf[MAVLINK_MAX_PACKET_LEN];
 	unsigned _network_buf_len;
 #endif
+
+	const char *_mavlink_wifi_name;
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.h]: Can you rename this to `_interface_name` and default to nullptr, so that the previous behavior is unchanged and `__DEFAULT_MAVLINK_WIFI` is not needed?

File: platforms/posix/src/main.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bkueng) [reply platforms/posix/src/main.cpp]: Weird. What are the individual `argv` strings and how do you run it? Does it actually do the redirect?

File: posix-configs/bbblue/px4.config (hunk 5 above)
[the remark this answers is not in the document]
Comment (bkueng) [reply posix-configs/bbblue/px4.config]: The land detector does not need specific HW support. You should be able to just let it run.

File: platforms/posix/src/main.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply platforms/posix/src/main.cpp]: Checked my porting notes. When run from command line, argv does not contain the redirect part. The above redirect handling was added when handling issues in px4 auto start during boot. This also brings up another issue encountered when setting up auto start on BeagleBone Blue. Here are some of my porting notes:

Notes for autostart using systemd:
( px4-*.service currently generated too much log and  increased disk usage quickly
  so use /etc/rc.local for now
  BTW, both methods are working except the above px4 problem
)

Scripts and comments in my /etc/rc.local:

#/home/debian/px4/px4  /home/debian/px4/posix-configs/px4.config > /home/debian/px4/PX4.log &
#
# After initial output, it continuosly fill up the log with the following garbage data:
#     pxh> ESC[2K^Mpxh> <FF>ESC[2K^Mpxh> <FF><FF>ESC[2K^Mpxh> <FF><FF><FF>ESC[2K^Mpxh> <FF><FF><FF><FF>ESC[2K^Mpxh> ......
# Disk usage can jump from about 50% to 80% quickly
# so have to redirect output to /dev/null for now
#
/home/debian/px4/px4  /home/debian/px4/posix-configs/px4.config  > /dev/null &

File: posix-configs/bbblue/px4.config (hunk 5 above)
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply posix-configs/bbblue/px4.config]: enabled it

File: src/modules/mavlink/mavlink_main.h (hunk 10 above)
Comment (UAV-Pilot) [reply src/modules/mavlink/mavlink_main.h]: yes, did so

File: src/modules/mavlink/mavlink_main.cpp [hunk 11]
```diff
@@ -3015,6 +3023,8 @@ Start mavlink on UDP port 14556 and enable the HIGHRES_IMU message with 50Hz:
 #endif
 	PRINT_MODULE_USAGE_PARAM_STRING('m', "normal", "custom|camera|onboard|osd|magic|config|iridium|minimal",
 					"Mode: sets default streams and rates", true);
+	PRINT_MODULE_USAGE_PARAM_STRING('n', "SoftAp", "SoftAp|wlan|name of custom wifi interface",
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.cpp]: Can you update this as well to:
`PRINT_MODULE_USAGE_PARAM_STRING('n', nullptr, "<interface_name>", "wifi/ethernet interface name", true); `

File: platforms/posix/src/main.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bkueng) [reply platforms/posix/src/main.cpp]: You need to start it with: `/home/debian/px4/px4 -d /home/debian/px4/posix-configs/px4.config > /home/debian/px4/PX4.log &`

> The above redirect handling was added when handling issues in px4 auto start during boot.

Does `strcmp` work as well? I just want to avoid introducing corner cases where the parsing is incorrect. Can you please double-check that it's still an issue?

File: src/modules/mavlink/mavlink_main.cpp (hunk 11 above)
Comment (UAV-Pilot) [reply src/modules/mavlink/mavlink_main.cpp]: Sorry I missed it. Updated.

File: platforms/posix/src/main.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (UAV-Pilot) [reply platforms/posix/src/main.cpp]: "-d" optiion worked and I removed the redirect flag and handled the case when test_data is absent. Previously when I used "-d /home/debian/px4" option, a link like the following was created, and caused failure when re-launching px4:

debian@beaglebone:~/px4$ ls -l
lrwxrwxrwx 1 root   root         26 Jul 31 19:02 test_data -> /home/debian/px4/test_data

I mistaken the whole of "-d /home/debian/px4" as data directory option and always avoided using the "-d" later.

The test_data directory is not small ( > 1MB )  and I do not think it's necessary to run px4, so I changed code not to create this link when the target directory does not exist.

**Comment by UAV-Pilot** [comment]
Thanks!

**Comment by UAV-Pilot** [comment]
updated submodule src/lib/DriverFramework

**Comment by dagar** [comment]
Driverframework update https://github.com/PX4/Firmware/pull/10141.
